tests/test_p4_revert_unshelved.py

```diff
@@ -131,3 +132,60 @@ def create_file(filename):
 
     for result, expected in zip(diff, expected_path):
         assert result == expected
+
+
+class RevertEnvironment(utils.TestEnvironment):
+    def __init__(self, directory, perforce_workspace):
+        self.work_dir = directory
+        self.vcs_cooking_dir = perforce_workspace.root_directory
+        self.repo_file = perforce_workspace.repo_file
+        self.nonwritable_file = perforce_workspace.nonwritable_file
+        self.p4 = perforce_workspace.p4
+        self.depot = perforce_workspace.depot
+        self.configs_file = directory.join("configs.py")
+
+        super(RevertEnvironment, self).__init__(directory, "main")
+
+
+@pytest.fixture()
+def perforce_environment(tmpdir, perforce_workspace):
+    yield RevertEnvironment(tmpdir, perforce_workspace)
+
+
+def test_p4_error_revert(perforce_environment, stdout_checker, capsys):
+    p4 = perforce_environment.p4
+    p4_file = perforce_environment.repo_file
+
+    settings = perforce_environment.settings
+    settings.Vcs.type = "p4"
+    settings.Launcher.output = "console"
+    settings.PerforceVcs.port = perforce_environment.p4.port
+    settings.PerforceVcs.user = perforce_environment.p4.user
+    settings.PerforceVcs.password = perforce_environment.p4.password
+    settings.PerforceMainVcs.client = "p4_disposable_workspace"
+    settings.PerforceMainVcs.force_clean = True
+    settings.PerforceWithMappings.project_depot_path = perforce_environment.depot
+
+    config = """
+from _universum.configuration_support import Variations
+
+configs = Variations([dict(name="Restrict changes", command=["chmod", "-R", "-w", "."]),
+                      dict(name="Check", command=["ls", "-la"])])
+""".format(p4_file.basename, p4_file.basename)
+    perforce_environment.configs_file.write(config)
+
+    # Prepare SHELVE_CHANGELIST
+    p4.run_edit(perforce_environment.depot)
+    p4_file.write("This is a change.\n")
+    change = p4.fetch_change()
+    change["Description"] = "CL for shelving"
+    shelve_cl = p4.save_change(change)[0].split()[1]
+    p4.run_shelve("-fc", shelve_cl)
+    settings.PerforceMainVcs.shelve_cls = [shelve_cl]
+
+    result = universum.run(perforce_environment.settings)
```

Comment:
The problem of using 'universum_runner' instead of all this mess (as in 'test_integration') is that when working in Docker, Universum is launched as root, and therefore Perforce (P4 CLI) in it is executed as root, and default 'revert' works on any files, even read-only, even in read-only directories